lib/matplotlib/backend_bases.py

```diff
@@ -820,10 +820,8 @@ def get_antialiased(self):
         return self._antialiased
 
     def get_capstyle(self):
-        """
-        Return the capstyle as a string in ('butt', 'round', 'projecting').
-        """
-        return self._capstyle
+        """Return the `.CapStyle`."""
+        return self._capstyle.name
```

Comment:
We have discussed this before, but my vote is that the minuscule breakage here of just returning `self._capstyle` is worth it.  I can't imagine anyone programatically wanting to call this...

However, you *may* want to implement an `__equals__` operator for these classes so `capstyle == 'round'` returns True.  I'm not clear from your mixing above whether that is already the case.  If so, then I really do not support keeping the type the same, since the odds of someone checking the type here are miniscule.

Reply:
This is a super-good point, and I even left in the commit in this PR (reverted by a later commit) which accomplishes this, in case we ever change our minds.

Since we use `str` as a mix-in, we get `__eq__`, `__add__`, etc, etc all for free from `str`. The long term goal is for by the end of GSOD (around March), I want to have *all* of the "styles" encapsulated in this same kind of new-style class (linestyle, hatch, joinstyle, capstyle, color, etc.). The original reason why this change was veto'd is that it seemed like a bad idea to introduce these new objects into our API if only one or two of them are complete (currently, only joinstyle/capstyle, although linestyle/hatch are *very* close to done).

The current plan (if I understand the latest state of the discussion correctly) is to put in all the classes, and then sometime around the end of GSOD when they're all started to "get in", make a separate PR that does all the API breaking at once, preferably with some loud announcements?

As I've brought up in the GSOD call, my initial impression is that this doesn't really make sense for something like `JoinStyle`, where the only API break is going to be if someone asks `type(js)`. However, @tacaswell is right to be dubious of this approach, as I do have to admit that even when very accomplished, careful developers migrate something from a string to a more advanced class, there tends to be very subtle ways in which this breaks stuff for users (I'm thinking for example of Python's move to `dict_keys` objects).

Reply:
Almost forgot. In the current state of the PR, there is one *real life* use case then in which the new class object can't be used directly in place of the old strings. Since `Enum` implements `__hash__`, then if a backend is using the `JoinStyle` to query a dictionary like:
```python
# taken from SVG backend verbatim
_capstyle_d = {'projecting': 'square', 'butt': 'butt', 'round': 'round'}
```
Then the following raises a `KeyError`. 
```python
_capstyle_d[CapStyle('projecting')]
```

This means that likely all custom backends that handle join or capstyles would be broken by this PR if we let it go in as-is.

However, this is easy enough to fix:
```python
# in the superclass
def __hash__(self):
    return str(self).__hash__()
```
will make `_capstyle_d[CapStyle('projecting')]` work again.